tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+
+  memcpy(index, &nonce_a_len, sizeof(size_t));
+  index += sizeof(size_t);
+  memcpy(index, nonce_a, nonce_a_len);
+  index += nonce_a_len;
+  memcpy(index, &nonce_b_len, sizeof(size_t));
+  index += sizeof(size_t);
+  memcpy(index, nonce_b, nonce_b_len);
+  index += nonce_b_len;
+  memcpy(index, &id_len, sizeof(size_t));
+  index += sizeof(size_t);
+  memcpy(index, id, id_len);
+
+  // Encrypt the nonce response and then clean up the unencrypted response.
+  int result =
+    encrypt_with_key_pair(ctx, message, message_len, response, response_len);
+
+  memset(message, 0, message_len);
+  free(message);
+
+  // Handle encryption errors if any occurred.
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to encrypt the nonce response.");
+    return 1;
+  }
+
+  return 0;
+}
+
+//
+// parse_nonce_response()
+//
+int parse_nonce_response(EVP_PKEY_CTX * ctx,
+                         unsigned char *response, size_t response_len,
+                         unsigned char **nonce_a, size_t *nonce_a_len,
+                         unsigned char **nonce_b, size_t *nonce_b_len,
+                         unsigned char **id, size_t *id_len)
+{
+  // Decrypt the nonce response.
+  unsigned char *message = NULL;
+  size_t message_len = 0;
+  int result =
+    decrypt_with_key_pair(ctx, response, response_len, &message, &message_len);
+
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to decrypt the nonce response.");
+    return 1;
+  }
+  unsigned char *index = message;
+
+  // Parse out the first nonce.
+  memcpy(nonce_a_len, index, sizeof(size_t));
+  index += sizeof(size_t);
+  *nonce_a = calloc(*nonce_a_len, sizeof(unsigned char));
+  if (*nonce_a == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the first nonce buffer.");
+
+    *nonce_a_len = 0;
+
+    memset(message, 0, message_len);
+    free(message);
+
+    return 1;
+  }
+  memcpy(*nonce_a, index, *nonce_a_len);
+  index += *nonce_a_len;
+
+  // Parse out the second nonce.
+  memcpy(nonce_b_len, index, sizeof(size_t));
+  index += sizeof(size_t);
+  *nonce_b = calloc(*nonce_b_len, sizeof(unsigned char));
+  if (*nonce_b == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the second nonce buffer.");
+
+    *nonce_b_len = 0;
+
+    memset(*nonce_a, 0, *nonce_a_len);
+    free(*nonce_a);
+    *nonce_a = NULL;
+    *nonce_a_len = 0;
+
+    memset(message, 0, message_len);
+    free(message);
+
+    return 1;
+  }
+  memcpy(*nonce_b, index, *nonce_b_len);
+  index += *nonce_b_len;
+
+  // Parse out the ID.
+  memcpy(id_len, index, sizeof(size_t));
+  index += sizeof(size_t);
+  *id = calloc(*id_len, sizeof(unsigned char));
+  if (*id == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the ID buffer.");
+
+    memset(*nonce_a, 0, *nonce_a_len);
+    free(*nonce_a);
+    *nonce_a = NULL;
+    *nonce_a_len = 0;
+
+    memset(*nonce_b, 0, *nonce_b_len);
+    free(*nonce_b);
+    *nonce_b = NULL;
+    *nonce_b_len = 0;
+
+    *id_len = 0;
+
+    memset(message, 0, message_len);
```

Comment:
We should probably use secure_memset() (in tpm2/src/util/memory_util.c), which prevents against potential compiler optimization issues.

Reply:
Done